### PR TITLE
Create CONTRIBUTING.md

### PR DESCRIPTION
**Pull request by philberty**
This contributing guide should answer frequently asked questions when people submit code to GCCRS.

### CODE REVIEW

**Review by flip1995 — COMMENTED** [review]
Some things that may be worth mentioning:

1. The PR policy: Everything has to go through a PR
2. Reviewers/Maintainers of the project to ping/ask for reviews/questions (aka people who have bors rights)

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+* If you're unable to find an open issue addressing the problem, [open a new one](https://github.com/Rust-GCC/gccrs/issues/new). Be sure to include a **title and clear description**, as much relevant information as possible, and a **code sample** or an **executable test case** demonstrating the expected behavior that is not occurring.
+
+#### **Did you write a patch that fixes a bug?**
+
+* Open a new GitHub pull request with the patch.
+
+* Ensure the PR description clearly describes the problem and solution. Include the relevant issue number if applicable.
+
+* Before submitting, GCCRS requires copyright assignment please read the [Contributing to GCC](https://gcc.gnu.org/contribute.html) guide to know more.
```
Comment (bjorn3) [inline CONTRIBUTING.md:15]:
```suggestion
* Before submitting, GCCRS requires copyright assignment. Please read the [Contributing to GCC](https://gcc.gnu.org/contribute.html) guide to know more.
```

**Comment by philberty** [comment]
Thanks for the suggestions they look good to me. The only caveat I would have is that we need to periodically merge with GCC which is a very big merge commit which I don't think should require it to be via PR but I am also not opposed to that either.

**Comment by flip1995** [comment]
> The only caveat I would have is that we need to periodically merge with GCC which is a very big merge commit which I don't think should require it to be via PR but I am also not opposed to that either.

Yes, you should also add this as the exception for this rule.

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -24,6 +24,13 @@
 
 * Ask any question about how to use GCCRS in [Zulip](https://gcc-rust.zulipchat.com/).
 
+### ** PR Policy **
+
+* The PR policy: Everything has to go through a PR
+  - An exception to this rule will be the merge commits of updating the repo against upstream GCC
+* Reviewers/Maintainers of the project (aka people who have bors rights) should be pinged for reviews/questions.
```
Comment (flip1995) [inline CONTRIBUTING.md:31]: Catch-22: The only people who can see this list are people who have bors rights.

**Comment by dkm** [comment]
Looks good !

**Review by dkm — APPROVED** [review]

**Review by SimplyTheOther — APPROVED** [review]
Looks good to me.

**Comment by NalaGinrut** [comment]
For the new features, it's better to have related tests.
Bugfix should have a test if any possible.

However, some cases are not suitable for auto testing, so I think this should be a suggestion.

**Comment by philberty** [comment]
Sounds good i will add a note about tests

**Review by NalaGinrut — APPROVED** [review]

**Comment by dkm** [comment]
Is this the correct place to ask for how the PR should look ? 
- a PR can have one or several commits (split should have a technical/logical reason, ie. no fixup-ish commit)
- no merge commit unless there's a good reason

TLDR: changes should be rebased over latest master branch and nicely split.

**Comment by flip1995** [comment]
I would say this is the perfect place. But to something like this, I would add an additional note, that this only applies when merging the PR, not while the PR is WIP. Otherwise people will start squash-force pushing which is a pain to review (I call this `git canpf` – `git commit --amend --no-edit && git push --force` and may or may not have an alias for it).

**Comment by dkm** [comment]
Ok, I'll try to add something about this. I must confess I'm guilty of push-forcing... I'll avoid this for future review :D

File: CONTRIBUTING.md (hunk 2 above)
Comment (dkm) [reply CONTRIBUTING.md:31]: I would add :
```
* Changes must be rebased (no merge commit) over latest `master` branch and be reasonably split
* During the review process, new changes should not be forced-pushed except for the last grooming step
```

Comment (flip1995) [reply CONTRIBUTING.md:31]: Force pushing isn't really the problem. If you rebase a PR on top of master you'll have to force push, but the commits still exist in the previous order. The problem is, if you only have one commit in a PR and every change is amended to that commit and then force pushed.

I would write the second point as 

```
* During the development/review process, changes to the PR should be pushed as new commits and only squashed together once the PR is ready to get merged.
```

**Comment by NalaGinrut** [comment]
For the merge commit issue, I think we need to make it flexible:
1. In normal cases there shouldn't be any merge commit. We should cooperate with rebasing. Because in this model, the conflicts can be resolved and `push -f` that will not affect the master.
2. However, the world is not perfect.  If there's a conflict when we behind the upstream many commits, the rebase mode requires the maintainer to fix the conflicts repeatedly. In this situation, the maintainer may choose a better way.

**Comment by philberty** [comment]
I've added some more to the PR policy based on comments here. Let's merge this and please feel free to raise issues/discussion around this. This is something to help and not hinder development.
